### ISSUE TITLE
ENH: Expose values of dual variables in milp/linprog

### ISSUE THREAD

**Issue by fuglede** [scipy/scipy#17951]
### Is your feature request related to a problem? Please describe.

The function `scipy.optimize._highs._highs_wrapper._highs_wrapper` provides information about the solution to the dual problem through the attributes `"lambda"` and `"margs_bnds"`, but that information is not carried through `scipy.optimize._milp.milp`; that is, in `scipy.optimize._milp.milp`, parts of what is in `highs_res` is not carried over to `res`.

Here's an example (taken [from here](https://github.com/lovasoa/highs-js)):

```
from scipy.optimize import milp
from scipy import optimize

c = np.array([1, 2, 4, 1])
A = np.array([[-1, 1, 1, 10], [1, -4, 1, 0], [0, 1, 0, -.5]])
lb = np.array([-np.inf, -np.inf, 0])
ub = np.array([20, 30, 0])
constraints = optimize.LinearConstraint(A=A, lb=lb, ub=ub)
bounds = optimize.Bounds([0, -np.inf, -np.inf, 2], [40, np.inf, np.inf, 3]) 
res = milp(c=-c, constraints=constraints, bounds=bounds)
```

Here, in the execution of `milp`, `highs_res` has the following value:

```
{'status': 7,
 'message': 'Optimal',
 'unscaled_status': 7,
 'x': [17.5, 1.0, 16.5, 2.0],
 'slack': [0.0, 0.0, 0.0],
 'lambda': [-1.5, -2.5, -10.5],
 'marg_bnds': <MemoryView of 'ndarray' at 0x7f4674241220>,
 'fun': -87.5,
 'simplex_nit': 2,
 'ipm_nit': 0,
 'crossover_nit': 0}
```

where `np.array(highs_res["marg_bnds"])` is

```
array([[0.  , 0.  , 0.  , 8.75],
       [0.  , 0.  , 0.  , 0.  ]])
```

and I would like to have this array, as well as `highs_res["lambda"]` (which is `[-1.5, -2.5, -10.5]`) available through the public interface.

This could be useful, for instance, for implementing branch and price solvers where the values of the dual variables of the reduced master problem is used to solve the pricing problem.

**Comment by mdhaber** [comment]
`linprog` does expose these when there are no integration constraints. See "marginals":
https://docs.scipy.org/doc/scipy/reference/optimize.linprog-highs.html#optimize-linprog-highs
`milp` is often for problems with integer constraints, in which case dual variables are not available.

**Comment by fuglede** [comment]
@mdhaber Ah, thanks, `linprog(...).ineqlin.marginals` is indeed what I want, so that's great; sorry for missing that.

One thing that threw me off just now is that when using integer constraints, `ineqlin.marginals` is still there, but is just the all-zero array; I imagine that could be a bit misleading, and that it'd be better to leave it out altogether, or at least document that behavior? Similarly, `highs_res["lambda"]` is a zero array, but that's less of an issue since it's not part of the public API.

**Comment by mdhaber** [comment]
I agree; it would be better to leave out irrelevant things like that when there are integer constraints. PR welcome, all!

**Comment by fuglede** [comment]
Any preference for whether we should leave out the meaningless keys entirely, or if we should keep the schema intact and use e.g. `None` for `highs_res["lambda"]`, treating it as an `Optional[np.ndarray]`? It looks like there was an attempt at something in between in https://github.com/scipy/scipy/blob/6c4a3f3551f9ee52a75c8c3999a57bed8ea67deb/scipy/optimize/_linprog_highs.py#L381-L390 which handles the case that `"lambda"` is missing (which then doesn't actually happen even if there are non-continuous variables), and which uses `None` for e.g. `ineqlin.marginals`.

**Comment by mdhaber** [comment]
> which then doesn't actually happen even if there are non-continuous variables

I'm guessing it did originally, and then HiGHS changed to return zero arrays. Let's follow what it intended to do.

I would prefer not to do that with `milp` right now. There is other work, e.g. class-based wrapper of HiGHS, that takes priority for me.
